core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+func (store *CouchStorage) StoreObject(metaData common.MetaData, data []byte, status string) ([]common.StoreDestinationStatus, common.SyncServiceError) {
+	id := getObjectCollectionID(metaData)
+	var dests []common.StoreDestinationStatus
+	var deletedDests []common.StoreDestinationStatus
+
+	if metaData.DestinationPolicy != nil {
+		metaData.DestinationPolicy.Timestamp = time.Now().UTC().UnixNano()
+	}
+
+	if status == common.NotReadyToSend || status == common.ReadyToSend {
+		// The object was receieved from a service, i.e. this node is the origin of the object:
+		// set its instance id and create destinations array
+		newID := store.getInstanceID()
+		metaData.InstanceID = newID
+		if data != nil && !metaData.NoData && !metaData.MetaOnly {
+			metaData.DataID = newID
+		}
+
+		var err error
+		dests, deletedDests, err = createDestinationsFromMeta(store, metaData)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	existingObject := &couchObject{}
+	if err := store.getOne(id, existingObject); err != nil {
+		if err != notFound {
+			return nil, &Error{fmt.Sprintf("Failed to retrieve object's status. Error: %s.", err)}
+		}
+		existingObject = nil
+	}
+
+	if existingObject != nil {
+		if (metaData.DestinationPolicy != nil && existingObject.MetaData.DestinationPolicy == nil) ||
+			(metaData.DestinationPolicy == nil && existingObject.MetaData.DestinationPolicy != nil) {
+			return nil, &common.InvalidRequest{Message: "Can't update the existence of Destination Policy"}
+		}
+
+		if metaData.MetaOnly {
+			metaData.DataID = existingObject.MetaData.DataID
+			metaData.ObjectSize = existingObject.MetaData.ObjectSize
+			metaData.ChunkSize = existingObject.MetaData.ChunkSize
+		}
+		if metaData.DestinationPolicy != nil {
+			dests = existingObject.Destinations
+		}
+	}
+
+	newObject := couchObject{ID: id, MetaData: metaData, Status: status, PolicyReceived: false,
+		RemainingConsumers: metaData.ExpectedConsumers,
+		RemainingReceivers: metaData.ExpectedConsumers, Destinations: dests}
+
+	// In case of existing object, check if it had attachment and add to newObject if present
+	// This is done only in case of metaOnly update
+	// otherwise updated attachment will be added in the next block
+	if existingObject != nil {
+		newObject.Rev = existingObject.Rev
+		if metaData.MetaOnly && data == nil {
+			attachment, err := store.getAttachment(id)
+			if err != nil {
+				if err != notFound {
+					return nil, err
+				}
+			} else {
+				attachments := &kivik.Attachments{id: attachment}
+				newObject.Attachments = attachments
+			}
+		}
+	}
+
+	// Add attachment to newObject for NoData=false
+	if !metaData.NoData && data != nil {
+		content := ioutil.NopCloser(bytes.NewReader(data))
+		defer content.Close()
+
+		attachment := &kivik.Attachment{ContentType: "application/octet-stream", Content: content}
+		attachments := &kivik.Attachments{id: attachment}
+		newObject.Attachments = attachments
+	}
```

Comment:
I think you need an else here to cover the case:  `!metaOnly && no data passed in`
In this case, you need to remove stored data. See [here (mongo)](https://github.com/open-horizon/edge-sync-service/blob/2f0846c094e9271160bdf8f22fa981d0e63de4a1/core/storage/mongoStorage.go#L308) and [here (boltDB)](https://github.com/open-horizon/edge-sync-service/blob/2f0846c094e9271160bdf8f22fa981d0e63de4a1/core/storage/boltStorage.go#L297)

Reply:
In this function, a fresh object - _newObject_ - is created irrespective of whether there is an update to an existing object or a new object needs to be stored. In the case of an existing object, I check if it had an attachment before and add it to _newObject_ if present. This is done only in the case of metaOnly update because otherwise directly the updated attachment will be added in the next block of code. Cases where existing data needs to be removed, are handled implicitly by not adding the existing attachment to _newObject_. 
The cases in this function were a bit tricky so to make sure that I don't miss any, I created a truth table that is present in the design document [here](https://www.notion.so/Design-Document-CouchDB-Support-for-CSS-b57884223f3e4b53a57db5a3065db7cc#4b64000245c344fd83d0e091ea4fdf87). It would be helpful if you could confirm these cases in case I missed any. I also added comments to the code about these cases, but if these are ambiguous, please let me know and I will modify them.

Reply:
Yea, that table looks correct to me. Thanks for the explanation